File: src/005.LinkedList/002.AddTwoNumbers/2.两数相加.js
/*
 * @lc app=leetcode.cn id=2 lang=javascript
 *
 * [2] 两数相加
 */

// @lc code=start
/**
 * Definition for singly-linked list.
 * function ListNode(val, next) {
 *     this.val = (val===undefined ? 0 : val)
 *     this.next = (next===undefined ? null : next)
 * }
 */
/**
 * @param {ListNode} l1
 * @param {ListNode} l2
 * @return {ListNode}
 */
var addTwoNumbers = function(l1, l2) {
  let carry = 0;
  let head, sumNode = null;

  while (l1 || l2) {
    let sum = (l1 && l1.val || 0) + (l2 && l2.val || 0) + carry;

    if (!head) {
      head = sumNode = new ListNode(sum % 10);
    } else {
      sumNode.next = new ListNode(sum % 10);
      // 节点重新赋值不要忘记
      sumNode = sumNode.next;
    }

    carry = Math.floor(sum / 10);

    if (l1) {
      l1 = l1.next;
    }
    if (l2) {
      l2 = l2.next;
    }
  }

  if (carry > 0) {
    sumNode.next = new ListNode(carry);
  }

  return head;
};
// @lc code=end

File: src/005.LinkedList/002.AddTwoNumbers/2.两数相加.go
/*
 * @lc app=leetcode.cn id=2 lang=golang
 *
 * [2] 两数相加
 */

// @lc code=start
/**
 * Definition for singly-linked list.
 * type ListNode struct {
 *     Val int
 *     Next *ListNode
 * }
 */
func addTwoNumbers(l1 *ListNode, l2 *ListNode) (head *ListNode) {
	carry := 0
	var sumNode *ListNode
	
	for l1 != nil || l2 != nil {
		n1, n2 := 0, 0
		if l1 != nil {
			n1 = l1.Val
			l1 = l1.Next
		}
		if l2 != nil {
			n2 = l2.Val
			l2 = l2.Next
		}

		sum := n1 + n2 + carry
		sum, carry = sum % 10, sum / 10

		if head == nil {
			head = &ListNode{ Val: sum }
			sumNode = head
		} else {
			sumNode.Next = &ListNode{ Val: sum }
			sumNode = sumNode.Next
		}
	}

	if carry > 0 {
		sumNode.Next = &ListNode{ Val: carry }
	}

	return
}
// @lc code=end

